retrieval/discovery/kubernetes/discovery.go

```diff
@@ -392,62 +405,144 @@ func (kd *Discovery) watchNodes(events chan interface{}, done <-chan struct{}, r
 }
 
 // watchServices watches services as they come & go.
-func (kd *Discovery) watchServices(events chan interface{}, done <-chan struct{}, retryInterval time.Duration) {
+func (kd *Discovery) startServiceWatch(events chan<- interface{}, done <-chan struct{}, retryInterval time.Duration) {
 	until(func() {
-		req, err := http.NewRequest("GET", servicesURL, nil)
+		// We use separate target groups for each discovered service so we'll need to clean up any if they've been deleted
+		// in Kubernetes while we couldn't connect - small chance of this, but worth dealing with.
+		existingServices := kd.services
+
+		// Reset the known services.
+		kd.services = map[string]map[string]*Service{}
+
+		services, resourceVersion, err := kd.getServices()
 		if err != nil {
-			log.Errorf("Failed to watch services: %s", err)
+			log.Errorf("Cannot initialize services collection: %s", err)
 			return
 		}
-		values := req.URL.Query()
-		values.Add("watch", "true")
-		values.Add("resourceVersion", kd.servicesResourceVersion)
-		req.URL.RawQuery = values.Encode()
 
-		res, err := kd.queryAPIServerReq(req)
-		if err != nil {
-			log.Errorf("Failed to watch services: %s", err)
+		// Now let's loop through the old services & see if they still exist in here
+		for oldNSName, oldNS := range existingServices {
+			if ns, ok := services[oldNSName]; !ok {
+				for _, service := range existingServices[oldNSName] {
+					events <- &serviceEvent{deleted, service}
+				}
+			} else {
+				for oldServiceName, oldService := range oldNS {
+					if _, ok := ns[oldServiceName]; !ok {
+						events <- &serviceEvent{deleted, oldService}
+					}
+				}
+			}
+		}
+
+		// Discard the existing services map for GC.
+		existingServices = nil
+
+		for _, ns := range services {
+			for _, service := range ns {
+				events <- &serviceEvent{added, service}
+			}
+		}
+
+		var wg sync.WaitGroup
+		wg.Add(2)
+
+		go kd.watchServices(resourceVersion, events, done, &wg)
```

Comment:
The `wg` is not really a relevant parameter to that function. Something like this would be more idiomatic I'd say.

```
go func() {
    kd.watchServices(resourceVersion, events, done)
    wg.Done()
}()
```